OsmAnd/src/net/osmand/plus/plugins/externalsensors/dialogs/SelectExternalDeviceFragment.java

```diff
@@ -0,0 +1,227 @@
+	}
+
+	@Override
+	public void onResume() {
+		super.onResume();
+		noBluetoothCard.setVisibility(plugin.isBlueToothEnabled() ? View.GONE : View.VISIBLE);
+		updatePairedSensorsListeners(true);
+		updatePairedSensorsList();
+	}
+
+	@Override
+	public void onPause() {
+		super.onPause();
+		updatePairedSensorsListeners(false);
+	}
+
+	private void updatePairedSensorsListeners(boolean add) {
+		List<AbstractDevice<?>> devices = plugin.getPairedDevices();
+		for (AbstractDevice<?> device : devices) {
+			if (add) {
+				device.addListener(this);
+			} else {
+				device.removeListener(this);
+			}
+		}
+	}
+
+	private void updatePairedSensorsList() {
+		List<AbstractDevice<?>> devices = plugin.getPairedDevices();
+		ArrayList<AbstractDevice<?>> filteredDevices = new ArrayList<>();
+		for (AbstractDevice<?> device : devices) {
+			if (isDeviceForWidgetFieldType(device)) {
+				filteredDevices.add(device);
+			}
+		}
+		if (devices.isEmpty()) {
+			emptyView.setVisibility(View.VISIBLE);
+			contentView.setVisibility(View.GONE);
+		} else {
+			requireActivity().runOnUiThread(() -> {
+				appBar.setExpanded(false);
+				devicesListAdapter.setItems(filteredDevices);
+				contentView.setVisibility(View.VISIBLE);
+				emptyView.setVisibility(View.GONE);
+			});
+		}
+	}
+
+	private boolean isDeviceForWidgetFieldType(AbstractDevice<?> device) {
+		for (AbstractSensor sensor : device.getSensors()) {
+			for (SensorWidgetDataFieldType type :
+					sensor.getSupportedWidgetDataFieldTypes()) {
+				if (type == widgetDataFieldType) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	public static void showInstance(@NonNull FragmentManager manager,
+	                                @NonNull SensorWidgetSettingFragment targetFragment,
+	                                @NonNull SensorWidgetDataFieldType fieldType,
+	                                @Nullable String selectedDeviceId) {
+		if (AndroidUtils.isFragmentCanBeAdded(manager, TAG)) {
+			SelectExternalDeviceFragment fragment = new SelectExternalDeviceFragment();
+			Bundle arguments = new Bundle();
+			arguments.putInt(WIDGET_TYPE_KEY, fieldType.ordinal());
+			arguments.putString(SELECTED_DEVICE_ID_KEY, selectedDeviceId);
+			fragment.setTargetFragment(targetFragment, 0);
+			fragment.setArguments(arguments);
+			fragment.setRetainInstance(true);
+			fragment.show(manager, TAG);
+		}
+	}
+
+	@Override
+	public void onDeviceConnect(@NonNull AbstractDevice<?> device, @NonNull DeviceConnectionResult result, @Nullable String error) {
+		app.runInUIThread(this::updatePairedSensorsList);
+	}
+
+	@Override
+	public void onDeviceDisconnect(@NonNull AbstractDevice<?> device) {
+		app.runInUIThread(this::updatePairedSensorsList);
+	}
+
+	@Override
+	public void onSensorData(@NonNull AbstractSensor sensor, @NonNull SensorData data) {
+	}
+
+	public void onDeviceSelected(@Nullable AbstractDevice<?> device) {
+		((SensorWidgetSettingFragment) getTargetFragment()).selectNewDevice(device);
```

Comment:
Check if fragment is available/instanceof, [example](https://github.com/osmandapp/OsmAnd/blob/47a19dc1880f4fae81f19d428df344ce9c3f19df/OsmAnd/src/net/osmand/plus/backup/ui/ClearTypesBottomSheet.java#L74-L77)